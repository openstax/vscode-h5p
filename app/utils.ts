import { FileInfo, Message } from './models';
import fsExtra from 'fs-extra';
import os from 'os';

// 2 GB = 2 * 1024 * 1024 * 1024 B
export const MAX_SIZE_IN_BYTES = 2147483648;

export function isFileInfoMessage(
  message: Message
): message is Message<FileInfo> {
  return message.type === 'FileInfo';
}

/**
 * This method will delete all temporary uploaded files from the request
 */
export async function clearTempFiles(
  req: Request & { files: any }
): Promise<void> {
  if (!req.files) {
    return;
  }

  await Promise.all(
    Object.keys(req.files).map((file) =>
      req.files[file].tempFilePath !== undefined &&
      req.files[file].tempFilePath !== ''
        ? fsExtra.remove(req.files[file].tempFilePath)
        : Promise.resolve()
    )
  );
}
