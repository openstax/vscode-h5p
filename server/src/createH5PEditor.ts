import { Cache, caching } from 'cache-manager';
import debug from 'debug';
import * as H5P from '@lumieducation/h5p-server';

/**
 * Create a H5PEditor object.
 * Which storage classes are used depends on the configuration values set in
 * the environment variables.re If you set no environment variables, the local
 * filesystem storage classes will be used.
 *
 * CONTENTSTORAGE=mongos3 Uses MongoDB/S3 backend for content storage
 * CONTENT_MONGO_COLLECTION Specifies the collection name for content storage
 * CONTENT_AWS_S3_BUCKET Specifies the bucket name for content storage
 * TEMPORARYSTORAGE=s3 Uses S3 backend for temporary file storage
 * TEMPORARY_AWS_S3_BUCKET Specifies the bucket name for temporary file storage
 *
 * Further environment variables to set up MongoDB and S3 can be found in
 * docs/packages/h5p-mongos3/mongo-s3-content-storage.md and docs/packages/h5p-mongos3/s3-temporary-file-storage.md!
 * @param config the configuration object
 * @param localLibraryPath a path in the local filesystem in which the H5P libraries (content types) are stored
 * @param localContentPath a path in the local filesystem in which H5P content will be stored (only necessary if you want to use the local filesystem content storage class)
 * @param localTemporaryPath a path in the local filesystem in which temporary files will be stored (only necessary if you want to use the local filesystem temporary file storage class).
 * @param translationCallback a function that is called to retrieve translations of keys in a certain language; the keys use the i18next format (e.g. namespace:key).
 * @returns a H5PEditor object
 */
export default async function createH5PEditor(
    config: H5P.IH5PConfig,
    localLibraryPath: string,
    localContentPath?: string,
    localTemporaryPath?: string,
    localContentUserDataPath?: string,
    translationCallback?: H5P.ITranslationFunction
): Promise<H5P.H5PEditor> {
    let cache: Cache;
    if (process.env.CACHE === 'in-memory') {
        debug('h5p-server')(
            `Using in memory cache for caching library storage.`
        );
        cache = caching({
            store: 'memory',
            ttl: 60 * 60 * 24,
            max: 2 ** 10
        });
    } else {
        debug('h5p-example')('Not using any cache for library storage');
        // using no cache
    }



    // Depending on the environment variables we use different implementations
    // of the storage interfaces.

    const h5pEditor = new H5P.H5PEditor(
        new H5P.cacheImplementations.CachedKeyValueStorage('kvcache', cache), // this is a general-purpose cache
        config,
        process.env.CACHE ? new H5P.cacheImplementations.CachedLibraryStorage( new H5P.fsImplementations.FileLibraryStorage(
          localLibraryPath
      ), cache)
            : new H5P.fsImplementations.FileLibraryStorage(
              localLibraryPath
          ),
        new H5P.fsImplementations.FileContentStorage(localContentPath),
        new H5P.fsImplementations.DirectoryTemporaryFileStorage( localTemporaryPath),
        translationCallback,
        undefined,
        {
            enableHubLocalization: true,
            enableLibraryNameLocalization: true,
            lockProvider: new H5P.SimpleLockProvider()
        }
    );

    return h5pEditor;
}
