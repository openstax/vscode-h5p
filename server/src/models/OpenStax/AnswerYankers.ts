export type Yanker = (content: any) => [unknown, unknown];
export type Unyanker = (publicData: any, privateData: any) => unknown;

function yankByKeys(content: any, keys: string[]): [unknown, unknown] {
  const publicData = {};
  const privateData = {};
  Object.entries(content).forEach(([k, v]) => {
    if (keys.includes(k)) {
      privateData[k] = v;
    } else {
      publicData[k] = v;
    }
  });
  return [publicData, privateData];
}

export const blanksYanker: Yanker = (content) => {
  return yankByKeys(content, ['questions']);
};

export const multiChoiceYanker: Yanker = (content) => {
  return yankByKeys(content, ['answers']);
};

export const questionSetYanker: Yanker = (content) => {
  const yankBySubtype = (q) => {
    const [libraryName, _version] = q.library.split(' ');
    switch (libraryName) {
      case 'H5P.MultiChoice':
        return multiChoiceYanker(q.params);
      case 'H5P.Blanks':
        return blanksYanker(q.params);
      case 'H5P.TrueFalse':
        return trueFalseYanker(q.params);
      default:
        throw new Error(`Library, "${libraryName}," is unsupported`);
    }
  };
  const privateData: unknown[] = [];
  const publicData = {
    ...content,
    // NOTE: Impure map
    questions: content.questions.map((q) => {
      const [pub, priv] = yankBySubtype(q);
      privateData.push(priv);
      return {
        ...q,
        params: pub,
      };
    }),
  };
  return [publicData, privateData];
};

export const questionSetMerge: Unyanker = (
  publicData: any,
  privateData: any
) => {
  const copy = { ...publicData };
  copy.questions.forEach((q, idx) => {
    q.params = shallowMerge(q.params, privateData[idx]);
  });
  return copy;
};

export const trueFalseYanker: Yanker = (content) => {
  return yankByKeys(content, ['correct']);
};

export const shallowMerge: Unyanker = (publicData: any, privateData: any) => ({
  ...publicData,
  ...privateData,
});
