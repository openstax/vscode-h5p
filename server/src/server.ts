/* --------------------------------------------------------------------------------------------
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 * ------------------------------------------------------------------------------------------ */
import { connect } from 'http2';
import {
  createConnection,
  ProposedFeatures,
  InitializeParams,
  InitializeResult,
} from 'vscode-languageserver/node';

import { prepareEnvironment, startH5P } from './createH5PServer';

// Create a connection for the server, using Node's IPC as a transport.
// Also include all preview / proposed LSP features.
const connection = createConnection(ProposedFeatures.all);

let hasConfigurationCapability = false;
let hasWorkspaceFolderCapability = false;
let hasDiagnosticRelatedInformationCapability = false;

connection.onInitialize((params: InitializeParams) => {
  connection.console.log('Initializing server');
  const capabilities = params.capabilities;

  // Does the client support the `workspace/configuration` request?
  // If not, we fall back using global settings.
  hasConfigurationCapability = !!(
    capabilities.workspace && !!capabilities.workspace.configuration
  );
  hasWorkspaceFolderCapability = !!(
    capabilities.workspace && !!capabilities.workspace.workspaceFolders
  );
  hasDiagnosticRelatedInformationCapability = !!(
    capabilities.textDocument &&
    capabilities.textDocument.publishDiagnostics &&
    capabilities.textDocument.publishDiagnostics.relatedInformation
  );

  const result: InitializeResult = {
    capabilities: {},
  };
  if (hasWorkspaceFolderCapability) {
    result.capabilities.workspace = {
      workspaceFolders: {
        supported: true,
      },
    };
  }
  return result;
});

connection.onInitialized(() => {
  const inner = async (): Promise<void> => {
    const currentWorkspaces =
      (await connection.workspace.getWorkspaceFolders()) ?? [];
    if (currentWorkspaces.length > 0) {
      console.log('Preparing environment for server');
      await prepareEnvironment().then(async (e) => {
        console.log('Environment prepared');
        await startH5P().then((e) => {
          console.log('Starting server');
        });
      });
    }
  };
  inner().catch((e) => {
    throw e;
  });
});

// Listen on the connection
connection.listen();
