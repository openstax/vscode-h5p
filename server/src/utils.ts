import * as os from 'os';
import decompress from 'decompress';
import fs from 'fs';
import express from 'express';
import { DOMParser } from '@xmldom/xmldom';
import * as xpath from 'xpath-ts';
import {
  assertTrue,
  assertValue,
  isFalsy,
  unwrap,
} from '../../common/src/utils';

export interface MergeOptions {
  customMerge?: (lhs: unknown, rhs: unknown, options?: MergeOptions) => unknown;
  arrayMerge?: (
    lhs: unknown[],
    rhs: unknown[],
    options?: MergeOptions,
  ) => unknown[];
}

/* istanbul ignore next (pure function that depends solely on express) */
export function createH5PRouter<EditRequestType, ContentRequestType>(
  onPlay: (req: any, res: any) => Promise<void>,
  onEdit: (req: EditRequestType, res: any) => Promise<void>,
  onNew: (req: ContentRequestType, res: any) => Promise<void>,
  onSave: (req: ContentRequestType, res: any) => Promise<void>,
  onDelete: (req: ContentRequestType, res: any) => Promise<void>,
  onFetch: (req: ContentRequestType, res: any) => Promise<void>,
): express.Router {
  const router = express.Router();

  router.get('/:contentId/play', onPlay);
  router.get('/:contentId/edit', onEdit);
  router.post('/', onNew);
  router.patch('/:contentId', onSave);
  router.delete('/:contentId', onDelete);
  router.get('/', onFetch);

  return router;
}

export function getIps(external: boolean = false): string[] {
  if (!external) return ['localhost'];
  const interfaces = os.networkInterfaces();
  return Object.values(interfaces).flatMap((devInts) =>
    assertValue(devInts)
      .filter((int: { internal: any }) => isFalsy(int.internal))
      .filter((int: { family: string }) => int.family === 'IPv4')
      .map((int: { address: any }) => int.address),
  );
}

export function fsRemove(folderPath: string) {
  fs.rmSync(folderPath, { recursive: true, force: true });
}

export async function extractArchive(
  path: string,
  destinationFolder: string,
  options?: {
    deleteArchive?: boolean;
    filesToExtract?: string[];
    decompressOptions?: decompress.DecompressOptions;
    verbose?: boolean;
  },
): Promise<void> {
  const {
    deleteArchive,
    filesToExtract,
    decompressOptions = { strip: 0 },
    verbose = true,
  } = options ?? {};
  console.log(`Extracting file ${path}`);
  try {
    const files = await decompress(path, destinationFolder, decompressOptions);
    if (filesToExtract !== undefined) {
      files
        .filter((file) => !filesToExtract.includes(file.path))
        .forEach((file) => fsRemove(file.path));
    }
    if (verbose) {
      (filesToExtract ?? files.map((file) => file.path)).forEach((name) => {
        console.log(`${destinationFolder}/${name}`);
      });
    }
    if (deleteArchive === true) {
      fsRemove(path);
    }
  } catch (e) {
    /* istanbul ignore next */
    console.error(e);
  }
}

export class ParseError extends Error {}

export function parseToDOM(xmlString: string, mimeType = 'application/xml') {
  const locator = { lineNumber: 0, columnNumber: 0 };
  /* istanbul ignore next */
  const cb = () => {
    const pos = {
      line: locator.lineNumber - 1,
      character: locator.columnNumber - 1,
    };
    throw new ParseError(`ParseError: ${JSON.stringify(pos)}`);
  };
  const p = new DOMParser({
    locator,
    errorHandler: {
      warning: console.warn,
      error: cb,
      fatalError: cb,
    },
  });
  const doc = p.parseFromString(xmlString, mimeType);
  return doc;
}

export function parseBooksXML(booksXmlPath: string): {
  booksRoot: string;
  pagesRoot: string;
  mediaRoot: string;
  privateRoot: string;
  publicRoot: string;
} {
  const doc = parseToDOM(fs.readFileSync(booksXmlPath, 'utf-8'));
  const select = xpath.useNamespaces({
    bk: 'https://openstax.org/namespaces/book-container',
  });
  const bookVars: Record<string, string> = Object.fromEntries(
    (select('//bk:var', doc) as Element[]).map((n) =>
      [n.getAttribute('name'), n.getAttribute('value')].map(unwrap).map((i) => {
        const trimmed = i.trim();
        if (trimmed.length === 0) {
          /* istanbul ignore next */
          throw new Error('Found empty var in books.xml');
        }
        return trimmed;
      }),
    ),
  );
  return {
    booksRoot: bookVars['BOOKS_ROOT'] ?? '/collections',
    pagesRoot: bookVars['PAGES_ROOT'] ?? '/modules',
    mediaRoot: bookVars['MEDIA_ROOT'] ?? '/media',
    privateRoot: bookVars['PRIVATE_ROOT'] ?? '/private',
    publicRoot: bookVars['PUBLIC_ROOT'] ?? '/interactives',
  };
}

export function mergeByIndex(
  lhs: unknown[],
  rhs: unknown[],
  options?: MergeOptions,
) {
  return Object.values(
    recursiveMerge(
      Object.fromEntries(Object.entries(lhs)),
      Object.fromEntries(Object.entries(rhs)),
      options,
    ) as object,
  );
}

export function defaultValueMerge(lhs: unknown, rhs: unknown) {
  return rhs ?? lhs;
}

export function recursiveMerge(
  lhs: unknown,
  rhs: unknown,
  options?: MergeOptions,
): unknown {
  const { arrayMerge = mergeByIndex, customMerge = defaultValueMerge } =
    options ?? {};
  if (
    typeof lhs === 'object' &&
    typeof rhs === 'object' &&
    lhs != null &&
    rhs != null
  ) {
    const isArrayL = Array.isArray(lhs);
    const isArrayR = Array.isArray(rhs);
    if (isArrayL || isArrayR) {
      assertTrue(isArrayL && isArrayR, 'Expected two arrays');
      return arrayMerge(lhs as unknown[], rhs as unknown[], options);
    } else {
      const keysL = Object.keys(lhs);
      const keysR = Object.keys(rhs);
      const sharedKeys = keysL.filter((k) => keysR.indexOf(k) !== -1);
      const merged = Object.fromEntries(
        sharedKeys.map((k) => [
          k,
          recursiveMerge(Reflect.get(lhs, k), Reflect.get(rhs, k), options),
        ]),
      );
      return { ...lhs, ...rhs, ...merged };
    }
  } else {
    return customMerge(lhs, rhs, options);
  }
}
