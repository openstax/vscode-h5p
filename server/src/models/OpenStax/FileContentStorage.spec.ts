import * as fsExtra from 'fs-extra';
import OSStorage, { fixNamespaces } from './FileContentStorage';
import mockfs from 'mock-fs';
import path from 'path';
import Config, { SupportedLibrary } from './config';
import { NetworkMetadata } from '../../../../common/src/types';
import { IContentMetadata, IUser } from '@lumieducation/h5p-server';
import { assertValue } from '../../../../common/src/utils';
import { yankByKeysFactory } from './AnswerYankers';

const VIRTUAL_ROOT = '/root';
const CONFIG = new Config(VIRTUAL_ROOT, 'interactives', 'private');
const PRIVATE_PATH = CONFIG.privateContentDirectory;
const TEST_LIBRARY = 'Test.Library';
const MOCK_H5P_BASE: IContentMetadata = {
  title: 'default title',
  mainLibrary: TEST_LIBRARY,
  language: 'U',
  license: '',
  embedTypes: ['iframe'],
  preloadedDependencies: [],
  defaultLanguage: '',
};
const MOCK_OSMETA_BASE: Partial<NetworkMetadata> = {};
const MOCK_CONTENT_BASE = { isSolutionPublic: true };

function dirToObj(base: string) {
  const dir: Record<string, any> = {};
  fsExtra.readdirSync(base, { withFileTypes: true }).forEach((dirent) => {
    const p = path.join(base, dirent.name);
    dir[dirent.name] = dirent.isDirectory()
      ? dirToObj(p)
      : fsExtra.readFileSync(p, { encoding: 'utf-8' });
  });
  return dir;
}

const tryGet = (target: object, p: string | symbol) =>
  assertValue(Reflect.get(target, p), `Property not found ${String(p)}`);

const createStorageHarness = () => {
  const mockTempStorage = new Proxy({} as unknown as any, {
    get(target, p) {
      switch (p) {
        case 'getFileStream':
          return (tmpName: string) => ({
            pipe(output: fsExtra.WriteStream) {
              output.write(`<insert contents of ${tmpName}>`);
              output.end();
              return output;
            },
          });
      }
      return tryGet(target, p);
    },
  });
  const storage = new Proxy(new OSStorage(CONFIG, mockTempStorage), {
    get(target, p) {
      switch (p) {
        case 'addContent':
          return (
            metadata: IContentMetadata,
            content: any,
            user: IUser,
            id?: string | undefined,
          ) => {
            const osMeta = {
              ...MOCK_OSMETA_BASE,
              ...(content?.osMeta ?? {}),
            };
            return target.addContent(
              { ...MOCK_H5P_BASE, ...metadata } as unknown as IContentMetadata,
              { ...MOCK_CONTENT_BASE, ...content, osMeta },
              { ...user } as unknown as IUser,
              id,
            );
          };
      }
      return tryGet(target, p);
    },
  });
  // Patch for testing without being coupled to config
  Reflect.set(
    storage,
    'assertLibrary',
    (): SupportedLibrary => ({
      yankAnswers: yankByKeysFactory('questions'),
      isSolutionPublic: (content) =>
        Reflect.get(content, 'isSolutionPublic') === true,
    }),
  );
  return storage;
};

describe('File Content Storage', () => {
  beforeEach(() => {
    mockfs({
      [VIRTUAL_ROOT]: {
        interactives: {
          '1': {
            'h5p.json': JSON.stringify({
              ...MOCK_H5P_BASE,
              title: 'this should be stored in folder 1',
            }),
            'content.json': JSON.stringify({}),
            'metadata.json': JSON.stringify({
              extra: 'Something extra',
            }),
          },
          'image-test': {
            media: {
              'detailed.png': '',
              '_questions.png': '',
              'unreferenced.png': '',
            },
            'metadata.json': JSON.stringify({
              attachments: ['media/unreferenced.png'],
            }),
          },
        },
      },
    });
  });
  afterEach(() => {
    mockfs.restore();
  });

  it('saves content as expected', async () => {
    const storage = createStorageHarness();

    // Should use nickname as ID
    expect(
      await storage.addContent(
        {
          title: 'this should be stored in folder 2',
        } as unknown as IContentMetadata,
        {
          osMeta: { nickname: '2' },
        },
        {} as unknown as IUser,
      ),
    ).toBe('2');

    // Should use id argument
    expect(
      await storage.addContent(
        {
          title: 'this should be stored in folder 1234',
        } as unknown as IContentMetadata,
        {
          osMeta: {},
        },
        {} as unknown as IUser,
        '1234',
      ),
    ).toBe('1234');

    // Modifying existing content should not erase extra metadata
    await storage.addContent(
      {
        title: 'this should be stored in folder 1',
      } as unknown as IContentMetadata,
      {
        osMeta: {
          nickname: '1',
          books: [{ name: 'stax-test', lo: ['1-2-3'] }],
        },
      },
      {} as unknown as IUser,
      '1',
    );
    expect(await storage.getOSMeta('1')).toMatchObject({
      extra: 'Something extra',
    });

    // It should try to move images from temporary storage
    await storage.addContent(
      {} as unknown as IContentMetadata,
      {
        osMeta: {
          nickname: 'image-test',
        },
        text: `<p>Fill in the missing words</p>
        <p><img data-filename="test.png" src="http://.../temp-files/images/not-used.png#tmp"/></p>
        `,
      },
      {} as unknown as IUser,
      'image-test',
    );

    // It should include attachments
    await storage.addContent(
      {} as unknown as IContentMetadata,
      {
        osMeta: {
          nickname: 'image-test',
          _additional_field: [
            {
              content: '<img src="media/detailed.png"/>',
              solution_type: 'detailed',
            },
          ],
          _questions: [
            {
              content: `<p><img src="media/_questions.png"></p>`,
            },
          ],
        },
        text: `<p>Fill in the missing words</p>
        <p><img src="media/test.png"/></p>
        `,
      },
      {} as unknown as IUser,
      'image-test',
    );

    const result = dirToObj(VIRTUAL_ROOT);
    mockfs.restore();
    expect(result).toMatchSnapshot();
  });
  it('throws an error if tmp items do not have a name', async () => {
    const storage = createStorageHarness();
    await expect(async () => {
      await storage.addContent(
        {} as unknown as IContentMetadata,
        {
          osMeta: {
            nickname: 'image-test',
          },
          text: `<p>Fill in the missing words</p>
          <p><img src="http://.../temp-files/images/not-used.png#tmp"/></p>
          `,
        },
        {} as unknown as IUser,
        'image-text',
      );
    }).rejects.toThrowError(/data-filename/);
  });
  it('converts invalid searchable values when loading', async () => {
    const storage = createStorageHarness();
    const id = '1234';
    await storage.addContent(
      {
        title: 1234 as any,
      } as unknown as IContentMetadata,
      {
        osMeta: {},
      },
      {} as unknown as IUser,
      id,
    );
    const loaded = await storage.getMetadata(id);
    expect(typeof loaded.title).toBe('string');
    expect(typeof loaded.mainLibrary).toBe('string');
  });
  ['1234', '12345'].forEach((id) => {
    const isSolutionPublic = id !== '1234';
    it(
      `${id}: ` +
        (!isSolutionPublic
          ? 'removes private solutions'
          : 'leaves public solutions'),
      async () => {
        const storage = createStorageHarness();
        const h5pContent = {
          fake: 'to make sure it is saved too',
          questions: [
            `<p>This should${
              !isSolutionPublic ? ' not ' : ' '
            }appear in the public folder</p>`,
          ],
          isSolutionPublic,
        };
        await storage.addContent(
          {
            title: 'this should be stored in folder ' + id,
          } as unknown as IContentMetadata,
          {
            ...h5pContent,
          },
          {} as unknown as IUser,
          id,
        );
        expect(await storage.getParameters(id)).toStrictEqual(h5pContent);
        const result = dirToObj(VIRTUAL_ROOT);
        mockfs.restore();
        expect(result).toMatchSnapshot();
      },
    );
  });
  it('cleans up orphaned files on error', async () => {
    const storage = createStorageHarness();
    storage.getOSMeta = jest.fn().mockRejectedValue(new Error('TEST'));
    await expect(async () => {
      await storage.addContent(
        {} as unknown as IContentMetadata,
        {},
        {} as unknown as IUser,
        '1234',
      );
    }).rejects.toThrowError('TEST');
    expect(await storage.contentExists('1234')).toBe(false);
  });
  it('does not allow duplicate ids', async () => {
    const storage = createStorageHarness();
    await expect(async () => {
      await storage.addContent(
        {
          title: 'this should be stored in folder 1',
        } as unknown as IContentMetadata,
        {
          osMeta: { nickname: '1' },
        },
        {} as unknown as IUser,
      );
    }).rejects.toThrowError(/.*duplicate.*/i);
    expect(await storage.contentExists('1234')).toBe(false);
  });
  // Id 0 breaks the player (reason unknown)
  it('does not allow id 0', async () => {
    const storage = createStorageHarness();
    expect(async () => {
      await storage.addContent(
        {} as unknown as IContentMetadata,
        { osMeta: { nickname: '0' } },
        {} as unknown as IUser,
      );
    }).rejects.toThrow(/.*cannot be 0.*/);
  });
  it('deletes private data when content is deleted', async () => {
    const storage = createStorageHarness();
    const id = '12345';
    const h5pContent = {
      fake: 'to make sure it is saved too',
      questions: ['<p>Anything</p>'],
    };
    await storage.addContent(
      {
        title: 'this should be stored in folder 123456',
      } as unknown as IContentMetadata,
      {
        ...h5pContent,
        isSolutionPublic: false,
      },
      {} as unknown as IUser,
      id,
    );

    expect(await storage.contentExists(id)).toBe(true);
    expect(fsExtra.existsSync(`${PRIVATE_PATH}/12345`)).toBe(true);

    await storage.deleteContent(id);
    expect(await storage.contentExists(id)).toBe(false);
    expect(fsExtra.existsSync(`${PRIVATE_PATH}/12345`)).toBe(false);
  });
  it('Searches for private and public files', async () => {
    mockfs({
      [VIRTUAL_ROOT]: {
        interactives: {
          '1': {
            media: {
              'b.png': '<contents of b.png>',
            },
          },
        },
        private: {
          interactives: {
            '1': {
              media: {
                'a.png': '<contents of a.png>',
              },
            },
          },
        },
      },
    });
    const storage = createStorageHarness();
    expect(await storage.getFileStats('1', 'media/a.png')).toBeDefined();
    expect(await storage.getFileStats('1', 'media/b.png')).toBeDefined();
    await expect(storage.getFileStats('1', 'media/c.png')).rejects.toThrowError(
      /content-file-missing/,
    );
  });
  it('Moves attachments when visibility changes', async () => {
    mockfs({
      [VIRTUAL_ROOT]: {
        interactives: {
          '1': {
            media: {},
          },
        },
        private: {
          interactives: {
            '1': {
              ['content.json']: JSON.stringify({
                text: '<img src="media/a.png"/>',
              }),
              media: {
                'a.png': '<contents of a.png>',
              },
            },
          },
        },
      },
    });
    const storage = createStorageHarness();
    const findFilePath = tryGet(storage, '_findFilePath').bind(storage);
    const image = 'media/a.png';
    const id = '1';
    const content = {
      questions: [`<img src="${image}"/>`],
      osMeta: {
        nickname: id,
      },
    };
    await storage.addContent(
      {} as unknown as IContentMetadata,
      {
        ...content,
        isSolutionPublic: false,
      },
      {} as unknown as IUser,
      id,
    );
    expect(await findFilePath(id, image)).toMatch(
      new RegExp(`${VIRTUAL_ROOT}/${CONFIG.privatePath}`),
    );
    await storage.addContent(
      {} as unknown as IContentMetadata,
      {
        ...content,
        isSolutionPublic: true,
      },
      {} as unknown as IUser,
      id,
    );
    expect(await findFilePath(id, image)).toMatch(
      new RegExp(`${VIRTUAL_ROOT}/${CONFIG.contentPath}`),
    );
  });
  it('correctly handles unreferenced images', async () => {
    mockfs({
      [VIRTUAL_ROOT]: {
        interactives: {
          '1': {
            media: {
              'a.png': '<contents of a.png>',
            },
          },
        },
        private: {
          interactives: {
            '1': {
              ['content.json']: JSON.stringify({
                text: '<img src="media/a.png"/>',
              }),
              media: {
                'a.png': '<contents of a.png>',
              },
            },
          },
        },
      },
    });
    const storage = createStorageHarness();
    const findFilePaths = tryGet(storage, '_findFilePaths').bind(storage);
    const image = 'media/a.png';
    const id = '1';
    const baseContent = {
      text: `<img src="${image}"/>`,
      questions: [`<img src="${image}"/>`],
      isSolutionPublic: false,
      osMeta: {
        nickname: id,
      },
    };
    let paths: Partial<Record<'public' | 'private', string>>;
    await storage.addContent(
      {} as unknown as IContentMetadata,
      baseContent,
      {} as unknown as IUser,
      id,
    );
    // Image should exist in both public and private storage
    paths = await findFilePaths(id, image);
    expect(paths.public).toBeDefined();
    expect(paths.private).toBeDefined();

    // WHEN: Solutions are made public
    await storage.addContent(
      {} as unknown as IContentMetadata,
      {
        ...baseContent,
        isSolutionPublic: true,
        osMeta: {
          ...baseContent.osMeta,
        },
      },
      {} as unknown as IUser,
      id,
    );
    // THEN: Only the public storage should contain the image
    paths = await findFilePaths(id, image);
    expect(paths.public).toBeDefined();
    expect(paths.private).not.toBeDefined();

    // WHEN: Solutions are made private again
    await storage.addContent(
      {} as unknown as IContentMetadata,
      baseContent,
      {} as unknown as IUser,
      id,
    );
    // THEN: Both public and private storage should contain the image
    paths = await findFilePaths(id, image);
    expect(paths.public).toBeDefined();
    expect(paths.private).toBeDefined();

    // WHEN: Image is no longer referenced in public content
    await storage.addContent(
      {} as unknown as IContentMetadata,
      {
        ...baseContent,
        text: '',
      },
      {} as unknown as IUser,
      id,
    );
    // THEN: Image is not in public storage
    paths = await findFilePaths(id, image);
    expect(paths.public).not.toBeDefined();
    expect(paths.private).toBeDefined();

    // WHEN: Image is no longer referenced in either location
    await storage.addContent(
      {} as unknown as IContentMetadata,
      {
        ...baseContent,
        questions: [''],
        text: '',
      },
      {} as unknown as IUser,
      id,
    );
    // THEN: Image is gone
    paths = await findFilePaths(id, image);
    expect(paths.public).not.toBeDefined();
    expect(paths.private).not.toBeDefined();
  });
  it('still works when the contentPath does not exist', async () => {
    mockfs({
      [VIRTUAL_ROOT]: {},
    });
    const storage = createStorageHarness();
    const expectedId = '1';
    expect(
      await storage.addContent(
        {
          title: `this should be stored in folder ${expectedId}`,
        } as unknown as IContentMetadata,
        {
          osMeta: { nickname: '1' },
        },
        {} as unknown as IUser,
      ),
    ).toBe('1');
  });

  describe('fixNamespaces', () => {
    it('adds mathml namespace', () => {
      const content = {
        x: [
          '<math t="1"><mi a="test">x</mi></math>',
          '<div>something</div>',
          '<math><mrow xmlns:t="http://example.com" t:b="x"><mi>y</mi></mrow>' +
            '</math>',
          '<math><mi>z</mi></math>',
          '<math xmlns="not-xhtml-namespace"><mi>w</mi></math>',
        ].join('\n'),
      };
      fixNamespaces(content);
      expect(content.x).toBe(
        [
          '<math t="1" xmlns="http://www.w3.org/1998/Math/MathML">' +
            '<mi a="test">x</mi></math>',
          '<div>something</div>',
          '<math xmlns="http://www.w3.org/1998/Math/MathML">' +
            '<mrow xmlns:t="http://example.com" t:b="x">' +
            '<mi>y</mi>' +
            '</mrow></math>',
          '<math xmlns="http://www.w3.org/1998/Math/MathML"><mi>z</mi></math>',
          '<math xmlns="not-xhtml-namespace"><mi>w</mi></math>',
        ].join('\n'),
      );
    });
  });
});
