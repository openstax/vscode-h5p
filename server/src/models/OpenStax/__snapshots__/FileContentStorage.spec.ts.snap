// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`File Content Storage 1234: removes private solutions 1`] = `
{
  "interactives": {
    "1": {
      "metadata.json": "{"extra":"Something extra","books":["should-not-appear-in-snapshot"]}",
    },
    "1234": {
      "content.json": "{
  "fake": "to make sure it is saved too"
}
",
      "h5p.json": "{
  "title": "this should be stored in folder 1234",
  "mainLibrary": "H5P.Blanks",
  "language": "U",
  "license": "",
  "embedTypes": [
    "iframe"
  ],
  "preloadedDependencies": [],
  "defaultLanguage": ""
}
",
      "metadata.json": "{
  "is-solution-public": "false"
}
",
    },
  },
  "private": {
    "interactives": {
      "1234": {
        "content.json": "{
  "questions": [
    "<p>This should not appear in the content.json</p>"
  ]
}
",
      },
    },
  },
}
`;

exports[`File Content Storage 12345: leaves public solutions 1`] = `
{
  "interactives": {
    "1": {
      "metadata.json": "{"extra":"Something extra","books":["should-not-appear-in-snapshot"]}",
    },
    "12345": {
      "content.json": "{
  "fake": "to make sure it is saved too",
  "questions": [
    "<p>This should  appear in the content.json</p>"
  ]
}
",
      "h5p.json": "{
  "title": "this should be stored in folder 12345",
  "mainLibrary": "H5P.Blanks",
  "language": "U",
  "license": "",
  "embedTypes": [
    "iframe"
  ],
  "preloadedDependencies": [],
  "defaultLanguage": ""
}
",
      "metadata.json": "{
  "is-solution-public": "true"
}
",
    },
  },
}
`;

exports[`File Content Storage saves content as expected 1`] = `
{
  "interactives": {
    "1": {
      "metadata.json": "{
  "extra": "Something extra",
  "books": [
    "meta-3"
  ]
}
",
    },
    "1234": {
      "content.json": "{}
",
      "h5p.json": "{
  "title": "this should be stored in folder 1234",
  "mainLibrary": "something",
  "language": "U",
  "license": "",
  "embedTypes": [
    "iframe"
  ],
  "preloadedDependencies": [],
  "defaultLanguage": ""
}
",
    },
    "2": {
      "content.json": "{
  "this": "could",
  "literally": "be",
  "anything": "because",
  "it": "is",
  "very": "loosely",
  "defined": true
}
",
      "h5p.json": "{
  "title": "this should be stored in folder 2",
  "mainLibrary": "something",
  "language": "U",
  "license": "",
  "embedTypes": [
    "iframe"
  ],
  "preloadedDependencies": [],
  "defaultLanguage": ""
}
",
      "metadata.json": "{
  "books": [
    "meta-1"
  ]
}
",
    },
    "3": {},
    "4": {
      "content.json": "{}
",
      "h5p.json": "{
  "title": "this should be stored in folder 3",
  "mainLibrary": "something",
  "language": "U",
  "license": "",
  "embedTypes": [
    "iframe"
  ],
  "preloadedDependencies": [],
  "defaultLanguage": ""
}
",
      "metadata.json": "{
  "books": [
    "meta-2"
  ]
}
",
    },
    "5": {},
  },
}
`;
